### PR TITLE
Display owned trains that were rusted in log (fixes #1575)

### LINKED ISSUE #1575: Show rusted trains in log
When an event is logged such as `Event: 4 trains rust`, list the companies and trains that were rusted

### PR DESCRIPTION
**Pull request by lordscarlet**
Log now displays:

```
-- Phase 6 (Operating Rounds: 3, Train Limit: 2, Available Tiles: Yellow, Green, Brown ) --
-- Event: 3 trains rust (The Depot x2, N&W x2, C&A x2) --
PRR buys a 6 train for $630 from The Depot
PRR skips buy companies
Player 2 operates B&O
```

image

### CODE REVIEW

File: lib/engine/phase.rb [hunk 1]
```diff
@@ -107,12 +108,13 @@ def rust_trains!(train, entity)
...
 
         rusted_trains << t.name
+        owners << t.owner.name
         entity.rusted_self = true if entity && entity == t.owner
         t.rust!
       end
 
       @log << "-- Event: #{obsolete_trains.uniq.join(', ')} trains are obsolete --" if obsolete_trains.any?
-      @log << "-- Event: #{rusted_trains.uniq.join(', ')} trains rust --" if rusted_trains.any?
+      @log << "-- Event: #{rusted_trains.uniq.join(', ')} trains rust (#{owners.group_by(&:itself).map{|c,t| "#{c} x#{t.count}"}.join(', ')}) --" if rusted_trains.any?
```
Comment (tobymao) [inline lib/engine/phase.rb:117]: owners = Hash.new(0)

owners[t.owner.name] += 1

owners.map { |name, count| "#{...}"...

Comment (tobymao) [reply lib/engine/phase.rb:117]: also your lines are too long and the formatting is off, you can find them locally with 

docker-compose exec rack rake

**Comment by perwestling** [comment]
Here is how it looks like in 18MS:
image

The game data for this is shown at the end of this comment.

Update of 18MS might be needed. Question is if this can be generalized so that the printout can be generalized?

{
  "status": "active",
  "actions": [
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 1,
      "company": "AGS",
      "price": 30
    },
    {
      "type": "bid",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 2,
      "company": "MC",
      "price": 60
    },
    {
      "type": "bid",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 3,
      "company": "BS",
      "price": 40
    },
    {
      "type": "par",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 4,
      "corporation": "GMO",
      "share_price": "70,1,2"
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 5,
      "shares": [
        "GMO_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 6,
      "shares": [
        "GMO_2"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 7,
      "shares": [
        "GMO_3"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 8,
      "shares": [
        "GMO_4"
      ]
    },
    {
      "type": "par",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 9,
      "corporation": "IC",
      "share_price": "70,1,2"
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 10,
      "shares": [
        "IC_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 11,
      "shares": [
        "IC_2"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 12,
      "shares": [
        "IC_3"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 13,
      "shares": [
        "IC_4"
      ]
    },
    {
      "type": "par",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 14,
      "corporation": "L&N",
      "share_price": "70,1,2"
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 15,
      "shares": [
        "L&N_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 16,
      "shares": [
        "L&N_2"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 17,
      "shares": [
        "L&N_3"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 18,
      "shares": [
        "L&N_4"
      ]
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 19
    },
    {
      "type": "pass",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 20
    },
    {
      "type": "pass",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 21
    },
    {
      "type": "lay_tile",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 22,
      "hex": "H6",
      "tile": "57-0",
      "rotation": 1
    },
    {
      "type": "lay_tile",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 23,
      "hex": "H8",
      "tile": "4-0",
      "rotation": 1
    },
    {
      "type": "buy_train",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 24,
      "train": "2+-0",
      "price": 80,
      "variant": "2+"
    },
    {
      "type": "buy_train",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 25,
      "train": "2+-1",
      "price": 80,
      "variant": "2+"
    },
    {
      "type": "pass",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 26
    },
    {
      "type": "lay_tile",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 27,
      "hex": "B2",
      "tile": "58-0",
      "rotation": 0
    },
    {
      "type": "lay_tile",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 28,
      "hex": "C1",
      "tile": "8-0",
      "rotation": 3
    },
    {
      "type": "buy_train",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 29,
      "train": "2+-2",
      "price": 80,
      "variant": "2+"
    },
    {
      "type": "buy_train",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 30,
      "train": "2+-3",
      "price": 80,
      "variant": "2+"
    },
    {
      "type": "pass",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 31
    },
    {
      "type": "lay_tile",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 32,
      "hex": "C9",
      "tile": "6-0",
      "rotation": 1
    },
    {
      "type": "lay_tile",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 33,
      "hex": "B10",
      "tile": "8-1",
      "rotation": 4
    },
    {
      "type": "buy_train",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 34,
      "train": "3+-0",
      "price": 180,
      "variant": "3+"
    },
    {
      "type": "buy_train",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 35,
      "train": "3+-1",
      "price": 180,
      "variant": "3+"
    },
    {
      "type": "pass",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 36
    },
    {
      "type": "lay_tile",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 37,
      "hex": "H4",
      "tile": "4-1",
      "rotation": 1
    },
    {
      "type": "run_routes",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 38,
      "routes": [
        {
          "train": "2+-0",
          "connections": [
            [
              "H6",
              "H8"
            ],
            [
              "H10",
              "H8"
            ]
          ]
        },
        {
          "train": "2+-1",
          "connections": [
            [
              "H6",
              "H4"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 39,
      "kind": "payout"
    },
    {
      "type": "buy_train",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 40,
      "train": "3+-2",
      "price": 180,
      "variant": "3+"
    },
    {
      "type": "pass",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 41
    },
    {
      "type": "lay_tile",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 42,
      "hex": "D2",
      "tile": "8-2",
      "rotation": 0
    },
    {
      "type": "place_token",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 43,
      "city": "E1-0-0",
      "slot": 1
    },
    {
      "type": "run_routes",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 44,
      "routes": [
        {
          "train": "2+-2",
          "connections": [
            [
              "A1",
              "B2"
            ]
          ]
        },
        {
          "train": "2+-3",
          "connections": [
            [
              "B2",
              "C1",
              "D2",
              "E1"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 45,
      "kind": "payout"
    },
    {
      "type": "buy_train",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 46,
      "train": "3+-3",
      "price": 180,
      "variant": "3+"
    },
    {
      "type": "pass",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 47
    },
    {
      "type": "buy_company",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 48,
      "company": "MC",
      "price": 30
    },
    {
      "type": "lay_tile",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 49,
      "hex": "C7",
      "tile": "6-1",
      "rotation": 4
    },
    {
      "type": "place_token",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 50,
      "city": "6-1-0",
      "slot": 0
    },
    {
      "type": "run_routes",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 51,
      "routes": [
        {
          "train": "3+-0",
          "connections": [
            [
              "C9",
              "B10",
              "B12"
            ]
          ]
        },
        {
          "train": "2+-4",
          "connections": [
            [
              "C9",
              "C7"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 52,
      "kind": "payout"
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 53
    },
    {
      "type": "pass",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 54
    },
    {
      "type": "pass",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 55
    },
    {
      "type": "lay_tile",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 56,
      "hex": "H2",
      "tile": "8-3",
      "rotation": 4
    },
    {
      "type": "pass",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 57
    },
    {
      "type": "undo",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 58
    },
    {
      "type": "place_token",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 59,
      "city": "I1-0-0",
      "slot": 0
    },
    {
      "type": "run_routes",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 60,
      "routes": [
        {
          "train": "2+-0",
          "connections": [
            [
              "H6",
              "H8"
            ],
            [
              "H10",
              "H8"
            ]
          ]
        },
        {
          "train": "2+-1",
          "connections": [
            [
              "I1"
            ]
          ]
        },
        {
          "train": "3+-2",
          "connections": [
            [
              "I1",
              "H2",
              "H4"
            ],
            [
              "H6",
              "H4"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 61,
      "kind": "payout"
    },
    {
      "type": "pass",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 62
    },
    {
      "type": "lay_tile",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 63,
      "hex": "F2",
      "tile": "8-4",
      "rotation": 0
    },
    {
      "type": "run_routes",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 64,
      "routes": [
        {
          "train": "2+-2",
          "connections": [
            [
              "A1",
              "B2"
            ]
          ]
        },
        {
          "train": "2+-3",
          "connections": [
            [
              "B2",
              "C1",
              "D2",
              "E1"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 65,
      "kind": "payout"
    },
    {
      "type": "pass",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 66
    },
    {
      "type": "lay_tile",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 67,
      "hex": "D6",
      "tile": "3-0",
      "rotation": 2
    },
    {
      "type": "run_routes",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 68,
      "routes": [
        {
          "train": "3+-0",
          "connections": [
            [
              "C9",
              "B10",
              "B12"
            ]
          ]
        },
        {
          "train": "2+-4",
          "connections": [
            [
              "C9",
              "C7"
            ]
          ]
        },
        {
          "train": "3+-1",
          "connections": [
            [
              "C7",
              "D6"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 69,
      "kind": "payout"
    },
    {
      "type": "lay_tile",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 70,
      "hex": "H2",
      "tile": "25-0",
      "rotation": 0
    },
    {
      "type": "undo",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 71
    },
    {
      "type": "lay_tile",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 72,
      "hex": "H2",
      "tile": "25-0",
      "rotation": 4
    },
    {
      "type": "run_routes",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 73,
      "routes": [
        {
          "train": "2+-0",
          "connections": [
            [
              "H6",
              "H8"
            ],
            [
              "H10",
              "H8"
            ]
          ]
        },
        {
          "train": "2+-1",
          "connections": [
            [
              "I1"
            ]
          ]
        },
        {
          "train": "3+-2",
          "connections": [
            [
              "I1",
              "H2",
              "H4"
            ],
            [
              "H6",
              "H4"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 74,
      "kind": "payout"
    },
    {
      "type": "pass",
      "entity": "GMO",
      "entity_type": "corporation",
      "id": 75
    },
    {
      "type": "lay_tile",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 76,
      "hex": "G1",
      "tile": "8-5",
      "rotation": 3
    },
    {
      "type": "run_routes",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 77,
      "routes": [
        {
          "train": "2+-2",
          "connections": [
            [
              "A1",
              "B2"
            ]
          ]
        },
        {
          "train": "2+-3",
          "connections": [
            [
              "B2",
              "C1",
              "D2",
              "E1"
            ]
          ]
        },
        {
          "train": "3+-3",
          "connections": [
            [
              "H6",
              "H4"
            ],
            [
              "H4",
              "H2",
              "G1",
              "F2",
              "E1"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 78,
      "kind": "payout"
    },
    {
      "type": "pass",
      "entity": "IC",
      "entity_type": "corporation",
      "id": 79
    },
    {
      "type": "lay_tile",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 80,
      "hex": "C5",
      "tile": "4-2",
      "rotation": 2
    },
    {
      "type": "run_routes",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 81,
      "routes": [
        {
          "train": "3+-0",
          "connections": [
            [
              "C9",
              "B10",
              "B12"
            ]
          ]
        },
        {
          "train": "2+-4",
          "connections": [
            [
              "C9",
              "C7"
            ]
          ]
        },
        {
          "train": "3+-1",
          "connections": [
            [
              "D6",
              "C5"
            ],
            [
              "C7",
              "D6"
            ]
          ]
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "L&N",
      "entity_type": "corporation",
      "id": 82,
      "kind": "payout"
    }
  ],
  "id": "hs_aqlpwltj_1598728578",
  "players": [
    {
      "name": "Player 1"
    },
    {
      "name": "Player 2"
    },
    {
      "name": "Player 3"
    }
  ],
  "title": "18MS",
  "description": "",
  "max_players": "3",
  "mode": "hotseat",
  "user": {
    "id": 0,
    "name": "You"
  },
  "created_at": "2020-08-29",
  "loaded": true,
  "result": {}
}

**Comment by lordscarlet** [comment]
@perwestling how do I create an 18MS game? Is there a branch with it, or a switch to load alpha games or something? 

image

**Comment by perwestling** [comment]
> @perwestling how do I create an 18MS game? Is there a branch with it, or a switch to load alpha games or something?
> 
> image

@lordscarlet If you go to create game, and add "?all" to the URL, you can choose from all games. Then you can create an 18MS game as a hotseat game. 
18MS has single round draft instead of auction.
OR1.2 is green phase.
After OR2.2 2+ trains rust. 
There is one private that gives a free 2 train when bought in.
When 2+ rust, the corporations get a salvage amount for each 2+ train that is rusted (except the free 2+ train).

I think the rust code is in g_18_ms.rb

**Comment by lordscarlet** [comment]
@perwestling I get an error with that json.

```We're sorry this game cannot be continued due to GameError: Route must contain token```

If I back out to the last valid it is in OR 2.1

**Comment by lordscarlet** [comment]
@perwestling 
> 18MS has single round draft instead of auction.
> OR1.2 is green phase.
> After OR2.2 2+ trains rust.
> There is one private that gives a free 2 train when bought in.
> When 2+ rust, the corporations get a salvage amount for each 2+ train that is rusted (except the free 2+ train).
> 
> I think the rust code is in g_18_ms.rb

So should there be anything different about the message, based on the salvage amount, or should the message stay the same but add it to the 18ms phase code?

**Comment by perwestling** [comment]
@lordscarlet I guess you can complete this change for the general code. Then I can implement a similar change for 18MS.

**Comment by lordscarlet** [comment]
@tobymao code should be formatted properly, and @perwestling said I should let him handle 18MS.

**Review by tobymao — APPROVED** [review]

**Pull request merged**